cloud/pkg/imageprepullcontroller/controller/downstream.go

```diff
@@ -0,0 +1,269 @@
+		return
+	}
+
+	// handle timeout: could not receive image prepull msg feedback from edge node
+	// send prepull timeout response message to upstream
+	go dc.handleImagePrepullJobTimeoutOnEachNode(node, imagePrepull.Name, imagePrepullTemplateInfo.TimeoutSecondsOnEachNode)
+
+	// update imagePrepullJob status to prepulling
+	status := &v1alpha1.ImagePrepullStatus{
+		NodeName: node,
+		State:    v1alpha1.Prepulling,
+	}
+	err = patchImagePrepullStatus(dc.crdClient, imagePrepull, status)
+	if err != nil {
+		klog.Errorf("Failed to mark imagePrepullJob prepulling status: %v", err)
+	}
+}
+
+// handleNodeUpgradeJobTimeout is used to handle the situation that cloud don't receive upgrade result
+// from edge node within the timeout period.
+// If so, the ImagePrepullJobState will update to timeout
+func (dc *DownstreamController) handleImagePrepullJobTimeoutOnEachNode(node, jobName string, timeoutSecondsEachNode *uint32) {
+	var timeout uint32 = 360
+	if timeoutSecondsEachNode != nil && *timeoutSecondsEachNode != 0 {
+		timeout = *timeoutSecondsEachNode
+	}
+
+	receiveFeedback := false
+
+	_ = wait.Poll(10*time.Second, time.Duration(timeout)*time.Second, func() (bool, error) {
+		cacheValue, ok := dc.imagePrepullJobManager.ImagePrepullMap.Load(jobName)
+		if !ok {
+			receiveFeedback = true
+			klog.Errorf("ImagePrepullJob %s is not exist", jobName)
+			return false, fmt.Errorf("imagePrepullJob %s is not exist", jobName)
+		}
+		imagePrepullValue := cacheValue.(*v1alpha1.ImagePrepullJob)
+		for _, statusValue := range imagePrepullValue.Status.Status {
+			if statusValue.NodeName == node && (statusValue.State == v1alpha1.PrepullSuccessful || statusValue.State == v1alpha1.PrepullFailed) {
+				receiveFeedback = true
+				return true, nil
+			}
+			break
+		}
+		return false, nil
+	})
+
+	if receiveFeedback {
+		return
+	}
+	klog.Errorf("TIMEOUT to receive image prepull %s response from edge node %s", jobName, node)
+
+	// construct timeout image prepull response and send it to upstream controller
+	responseResource := buildPrepullResource(jobName, node)
+	resp := commontypes.ImagePrepullJobResponse{
+		NodeName: node,
+		State:    v1alpha1.PrepullFailed,
+		Message:  "timeout to receive response from edge",
+	}
+
+	respMsg := model.NewMessage("").
+		BuildRouter(modules.ImagePrepullControllerModuleName, modules.ImagePrepullControllerModuleGroup, responseResource, ImagePrepull).
+		FillBody(resp)
+	beehiveContext.Send(modules.ImagePrepullControllerModuleName, *respMsg)
+}
+
+// NewDownstreamController new downstream controller to process downstream imageprepull msg to edge nodes.
+func NewDownstreamController(crdInformerFactory crdinformers.SharedInformerFactory) (*DownstreamController, error) {
+	imagePrepullJobManager, err := manager.NewImagePrepullJobManager(crdInformerFactory.Operations().V1alpha1().ImagePrepullJobs().Informer())
+	if err != nil {
+		klog.Warningf("Create ImagePrepullJob manager failed with error: %s", err)
```

Comment:
Warnning logs are generally used for errros that can be ignored. Is this suitable?

Reply:
> Warnning logs are generally used for errros that can be ignored. Is this suitable?

This err will be return and logged at the outer layer. Possibly for the sake of concise error logs, other modules in cloudcore  all use `waning` to record logs.  I think it's enough. If it still need to be modified, we can track it in an issue and make the modification in all modules.

Reply:
Usually an error log print and returns the original error will not help us locate the error, my suggestion is to returns a wrapped error stack, like:
```golang
return fmt.Errorf("error message, err: %v", err)
// or github.com/pkg/error
return errors.WithMessage(err, "error message")
```
We can create a new issue to track and discuss.